build_tools/aws-sdk-code-generator/lib/aws-sdk-code-generator/endpoint_parameter.rb

```diff
@@ -0,0 +1,200 @@
+      # Built-in binding default values
+      # @retyrn [value, source].  source may be one of [operation, config, default]
+      def endpoint_parameter_value(service, param_name, param_data, operation=nil)
+        unless operation.nil?
+          value, source = [
+            static_context_param(operation, param_name), 'operation'
+          ]
+          value, source = [
+            context_param_value(service, operation, param_name), 'operation'
+          ] unless value
+          value, source = [
+            operation_context_param_value(operation, param_name), 'operation'
+          ] unless value
+        end
+
+        value, source = [
+          client_context_param_value(service, param_name, param_data),
+          'config'
+        ] unless value
+
+
+        # built-ins may be sourced from operation context in some cases
+        unless value
+          value, source = built_in_param_value(service, param_data)
+        end
+
+        [value || 'nil', source || 'default']
+      end
+
+      def client_context_param_value(service, param_name, param_data)
+        if service.api['clientContextParams']&.key?(param_name) &&
+          !param_data['builtIn']
+          "config.#{Underscore.underscore(param_name)}"
+        end
+      end
+
+      # built-ins may be sourced from operation context in some cases
+      def built_in_param_value(service, param_data)
+        source = 'config'
+        value =
+          case param_data['builtIn']
+          when 'AWS::Region'
+            'config.region'
+          when 'AWS::UseFIPS'
+            'config.use_fips_endpoint'
+          when 'AWS::UseDualStack'
+            if service.name == 'S3' || service.name == 'S3Control'
+              source = 'operation'
+              'context[:use_dualstack_endpoint]'
+            else
+              'config.use_dualstack_endpoint'
+            end
+          when 'AWS::Auth::AccountId'
+            'config.credentials.credentials.account_id'
+          when 'AWS::Auth::AccountIdEndpointMode'
+            'config.account_id_endpoint_mode'
+          when 'AWS::STS::UseGlobalEndpoint'
+            "config.sts_regional_endpoints == 'legacy'"
+          when 'AWS::S3::UseGlobalEndpoint'
+            "config.s3_us_east_1_regional_endpoint == 'legacy'"
+          when 'AWS::S3::Accelerate'
+            if service.name == 'S3' || service.name == 'S3Control'
+              source = 'operation'
+              'context[:use_accelerate_endpoint]'
+            else
+              'config.use_accelerate_endpoint'
+            end
+          when 'AWS::S3::ForcePathStyle'
+            'config.force_path_style'
+          when 'AWS::S3::UseArnRegion', 'AWS::S3Control::UseArnRegion'
+            'config.s3_use_arn_region'
+          when 'AWS::S3::DisableMultiRegionAccessPoints'
+            'config.s3_disable_multiregion_access_points'
+          when 'SDK::Endpoint'
+            'config.regional_endpoint ? nil : config.endpoint.to_s'
```

Comment:
This can be simplified to "(config.endpoint.to_s if config.regional_endpoint)"

Reply:
not sure its really a simplification, but changed.